### PR TITLE
Add cpu constraint

### PR DESCRIPTION
**Pull request by nicholasjackson**
Changed deployment options to adding ability to add CPU based constraints for a pod.

[https://github.com/openfaas/faas-netes/issues/103](https://github.com/openfaas/faas-netes/issues/103)

## Description
deploy.go has been changed adding the additional constraints to be set when a deployment api request is created.

## Motivation and Context
CPU and Memory constraints should be supported
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Tested locally with MiniKube, the image https://hub.docker.com/r/nicholasjackson/faas-netesd/ can be used for testing this change.

*Example config*
```yaml
provider:
  name: faas
  gateway: http://192.168.99.100:31112

functions:
  facedetect:
    lang: go-opencv
    handler: ./facedetect
    image: nicholasjackson/func_facedetect
    limits:
      memory: 10Mi
      cpu: 100m
```
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Nic, firstly thank you for agreeing to work on this and for raising the PR. Have you covered both update and create in the changes?

**Comment by nicholasjackson** [comment]
@alexellis good call, I am a k8s noob, I might have missed this, let me double check

**Comment by nicholasjackson** [comment]
@alexellis I am not seeing in the code where the constraints are updated at all should this be a feature of the update handler?  If so I can add this to the code.

File: handlers/deploy.go [hunk 1]
```diff
@@ -306,6 +307,22 @@ func createResources(request requests.CreateFunctionRequest) (*apiv1.ResourceReq
 		resources.Requests[apiv1.ResourceMemory] = qty
 	}
 
+	// Set CPU limits
+	if request.Limits != nil && len(request.Limits.CPU) > 0 {
```
[the remark this answers is not in the document]
Comment (nicholasjackson) [reply handlers/deploy.go]: Done, good suggestion

**Comment by alexellis** [comment]
Yes in the update handler - that's what's used for a rolling update.

**Comment by nicholasjackson** [comment]
The updates should be covered as the update handler uses the function `createResources` 
[https://github.com/nicholasjackson/faas-netes/blob/a785730d4039064be90b46d19a774507b0ca74ce/handlers/update.go#L64](https://github.com/nicholasjackson/faas-netes/blob/a785730d4039064be90b46d19a774507b0ca74ce/handlers/update.go#L64)

This is defined in the deploy handler. [https://github.com/nicholasjackson/faas-netes/blob/a785730d4039064be90b46d19a774507b0ca74ce/handlers/deploy.go#L288](https://github.com/nicholasjackson/faas-netes/blob/a785730d4039064be90b46d19a774507b0ca74ce/handlers/deploy.go#L288) and is the same function used when deploying a new function.

**Comment by alexellis** [comment]
Thanks. I'll review it this week.

**Comment by alexellis** [comment]
:100: Looks good to me. Can you find someone to verify / test in #contributors?

**Comment by stefanprodan** [comment]
I've tested this on GKE but it doesn't work:

```bash
kubectl -n openfaas set image deploy/faas-netesd faas-netesd=nicholasjackson/faas-netesd:latest
deployment "faas-netesd" image updated
```

```yaml
provider:
  name: faas
  gateway: http://localhost:8080

functions:
  promql:
    handler: ./promq
    image: stefanprodan/openfaas-promq
    labels:
      com.openfaas.scale.min: "2"
      com.openfaas.scale.max: "15"
    limits:
      cpu: 80m
      memory: 128Mi
    requests:
      cpu: "10m"
      memory: "64Mi"
```

```bash
faas-cli deploy -f ./promql.yml --gateway https://openfaas.weavedx.com
Deploying: promql.
No existing function to remove
Deployed.
```

Only the memory setting is there:

```bash
kubectl -n openfaas-fn get deployment/promql -o=yaml

        name: promql
        ports:
        - containerPort: 8080
          protocol: TCP
        resources:
          limits:
            memory: 128Mi
          requests:
            memory: 64Mi
```

```
faas-cli version

Commit: d1d38e9b2d5600a3485442b75641bf73b566313b
Version: 0.5.1

```

**Comment by stefanprodan** [comment]
Looks like the latest CLI release doesn't contain the CPU limits. Using CLI master branch it works:

```bash
./faas-cli deploy -f ./promql.yml --gateway https://openfaas.weavedx.com
Deploying: promql.
Function promql already exists, attempting rolling-update.
```

```bash
kubectl -n openfaas-fn get deployment/promql -o=yaml

        name: promql
        ports:
        - containerPort: 8080
          protocol: TCP
        resources:
          limits:
            cpu: 80m
            memory: 128Mi
          requests:
            cpu: 10m
            memory: 64Mi
```

**Comment by alexellis** [comment]
Great, thank you Stefan.

**Comment by nicholasjackson** [comment]
@stefanprodan aha I thought this was the case, just spun up a k8s cluster, to test this and report back and too late :(

Ah well at least I have some config to create a k8s cluster now

**Pull request merged**